teamcity/datadog_checks/teamcity/common.py

```diff
@@ -0,0 +1,152 @@
+# (C) Datadog, Inc. 2022-present
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+import re
+import time
+from collections import OrderedDict
+from copy import deepcopy
+
+from six import PY2
+
+if PY2:
+    from urlparse import urlparse
+else:
+    from urllib.parse import urlparse
+
+import requests
+
+from .constants import BUILD_EVENT, DEPLOYMENT_EVENT, RESOURCE_URL_MAP, STATUS_MAP
+
+
+def match(item, pattern_model):
+    if isinstance(pattern_model, str):
+        return re.search(pattern_model, item)
+    if isinstance(pattern_model, dict) and len(pattern_model) == 1:
+        return re.search(list(pattern_model.keys())[0], item)
+    elif isinstance(pattern_model, list):
+        return any(match(item, pattern) for pattern in pattern_model)
+    return False
+
+
+def filter_list(items, include_patterns, exclude_patterns):
+    excluded_items = [item for item in items if match(item, exclude_patterns) if exclude_patterns]
+    if not include_patterns:
+        return {item: None for item in items if item not in excluded_items}
+    else:
+        filtered_items = {}
+        for include_pattern in include_patterns:
+            filtered_items.update(
+                {item: include_pattern for item in items if item not in excluded_items and match(item, include_pattern)}
```

Comment:
I tried storing the `[item for items if item not in excluded_items]` generator as a variable and it works for the `if not include patterns` block, but not in the `for include_pattern in include_patterns:` block because I need to access the `item` variable to perform the `match` operation.  

I also tried using `itertools.product` for `filtered_items`, and I couldn't quite get it right because `included_items` is unknown and it's the thing that we're trying to obtain (included_items == filtered_items). I'd love to see examples of how you use in in your code in the future though!

Reply:
postponing to some later date